include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -739,44 +765,36 @@ class discard_iterator
     bool
     operator<(const discard_iterator& __it) const
     {
-        return __my_position_ - __it.__my_position_ < 0;
+        return *this - __it < 0;
     }
     bool
     operator>(const discard_iterator& __it) const
     {
-        return __my_position_ - __it.__my_position_ > 0;
-    }
-
-    difference_type
-    operator-(const discard_iterator& __it) const
-    {
-        return __my_position_ - __it.__my_position_;
+        return __it < *this;
     }
 
     discard_iterator&
     operator++()
     {
-        ++__my_position_;
-        return *this;
+        return *this += 1;
```

Comment:
If we want (and I think we do) to support all iterators categories we also need to do not add more requirements on the type `T` besides necessary ones. `operator++` is the requirement for every iterator, `operator+=` is for `random_access_iterator` only. I would suggest to do make our `operator++` and `operator--` more restrictive.

Need to closely look at other methods, probably I missed something.

Reply:
For the discard_iterator there's no underlying iterator type to support.  `__my_position_` is an integer so using `operator+=` here doesn't add more requirements to type `T` because the type isn't present for this iterator.

Reply:
For `discard_iterator` there is no type `T`, so using `operator+=` here isn't adding requirements.  There is no underlying iterator to support in this case.